contracts/src/ethereum/TokenBridge.sol

```diff
@@ -143,17 +143,34 @@ contract TokenBridge is
         messagingContract.consumeMessageFromL2(l2TokenBridge, payload);
     }
 
-    function deposit(address l1Token_, uint256 l2Recipient, uint256 amount) isApprovedToken(l1Token_) external {
-        IStaticATokenLM l1Token = IStaticATokenLM(l1Token_);
+    function deposit(IERC20 l1Token, uint256 l2Recipient, uint256 amount) external {
         l1Token.transferFrom(msg.sender, address(this), amount);
-        sendMessage(l1Token_, l2Recipient, amount);
+        sendMessage(address(l1Token), l2Recipient, amount);
     }
-    
-    function withdraw(address l1Token_, address recipient, uint256 amount) isApprovedToken(l1Token_) external {
-        consumeMessage(l1Token_, recipient, amount);
-        require(recipient != address(0x0), "INVALID_RECIPIENT");
-        IStaticATokenLM l1Token = IStaticATokenLM(l1Token_);
-        require(l1Token.balanceOf(msg.sender) - amount <= l1Token.balanceOf(msg.sender), "UNDERFLOW");
+
+    function depositUnderlying(
+        IStaticATokenLM l1Token,
+        uint256 l2Recipient,
+        uint256 amount,
+        uint16 refferalCode,
+        bool fromAsset
+    ) external {
+        amount = l1Token.deposit(msg.sender, amount, refferalCode, fromAsset);
+        this.deposit(l1Token, l2Recipient, amount);
```

Comment:
We should probably turn `deposit` into a `public` function and call it here without `this.f()` pattern. Calling an external function like that is really expensive, see this [SO post](https://ethereum.stackexchange.com/questions/19380/external-vs-public-best-practices).

Reply:
I've done it on Andrey's repo for the PR about StaticAToken, and I should probably change that.